### ISSUE TITLE
Add support of parsing ASTs in parallel

### ISSUE THREAD

**Issue by sabaka** [checkstyle/checkstyle#4409]
After investigation @soon found, that most time consuming part of checkstyle was AST parsing. We have to launch it in parallel after #4370 finished.

**Comment by rnveach** [comment]
How are we going to make any `TreeWalker` parallel outside of the rest of the process?
`TreeWalker` only accepts 1 file, and once `Checker.processFile` finishes, any messages still being generated in `TreeWalker` will be lost.

**Comment by soon** [comment]
Shouldn't we, instead, firstly implement multi thread support for `Checker` class, which will pick a file and process all file checks inside a separate thread?

**Comment by soon** [comment]
The current implementation of `LocalizedMessages` uses `TreeSet` to hold the messages.

However, `LocalizedMessage` defines `compareTo` as follows:

    @Override
    public int compareTo(LocalizedMessage other) {
        int result = Integer.compare(lineNo, other.lineNo);

        if (lineNo == other.lineNo) {
            if (columnNo == other.columnNo) {
                result = getMessage().compareTo(other.getMessage());
            }
            else {
                result = Integer.compare(columnNo, other.columnNo);
            }
        }
        return result;
    }

This is inconsistent with the `equals` method, which compares all fields (see [this test](https://github.com/soon/checkstyle/commit/0920895b1b9833092685115f45f0ced144a10604#diff-98d674b6093bb609c8772c7af45db711R41)).

I'm facing an issue with the `LocalizedMessages` while migrating some checks. For example, both `WhitespaceAround`  and `WhitespaceAfter` could produce the following checks:

> [ERROR] /home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/grammars/comments/InputFullOfSinglelineComments.java:31:4: 'for' is not followed by whitespace. [WhitespaceAround]

The only difference is the class name at the end of the line.

In ST mode this works pretty fine - either `WhitespaceAround` or `WhitespaceAfter` writes message to log and all further log messages will be ignored. This does not change between checkstyle calls.

In MT mode it is impossible to predict, will `WhitespaceAround` or `WhitespaceAfter` write the first log message, therefore the output may change between checkstyle calls.

So, my question is - should we replace TreeSet with HashSet and sort messages when all checks are finished, or should we improve `compareTo` method?

@rnveach @sabaka @romani

**Comment by soon** [comment]
Also, I'd like to discuss the `JavadocPackage` check. It prints error messages as follows:

> [ERROR] /home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4861blockcommentstyle/InputCommentsIndentationSurroundingCode.java:0: Missing package-info.java file. [JavadocPackage]

However, this message is the same as the following message:

> [ERROR] /home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4861blockcommentstyle/InputCommentsIndentationInSwitchBlock.java:0: Missing package-info.java file. [JavadocPackage]

Note, that these messages complains about the same directory. The only different part is the file name (`InputCommentsIndentationSurroundingCode.java` and `InputCommentsIndentationInSwitchBlock.java`). 

In ST mode it does not matter, because each call will produce same output. 

In MT mode file names may differ (only file names, not the directories).

Shouldn't we omit the file name and print the directory name only?

**Comment by MEZk** [comment]
> In ST mode it does not matter, because each call will produce same output.

How? You said that 

> that these messages complains about the same directory. The only different part is the file name 

This aspect is shown in your examples.

> In MT mode file names may differ (only file names, not the directories).

Same is for ST mode. Please explain.

**Comment by soon** [comment]
> > In ST mode it does not matter, because each call will produce same output.
>
> How? You said that
> 
> > that these messages complains about the same directory. The only different part is the file name
>
> This aspect is shown in your examples.
> 
> > In MT mode file names may differ (only file names, not the directories).
>
> Same is for ST mode. Please explain.
> 

@MEZk In ST mode each checkstyle run with same arguments produces the same output. This is because `JavadocPackage` check holds internal set of checked directories. It takes one file, verifies the file directory and then no more files from the same directory will be processed. Checkstyle processes all files one by one with the same order, therefore output does not change between checkstyle calls.

However, in MT mode the file order may be different. This is why in the first call it complains about `InputCommentsIndentationSurroundingCode` file and in the second call it complains about `InputCommentsIndentationInSwitchBlock` file.

**Comment by sabaka** [comment]
As for me, it would be better to complain about directory instead of file.

**Comment by rnveach** [comment]
> in MT mode the file order may be different

@soon See Issue #3340 about comments me and romani had that I wrote up on a similar issue.

Since we don't care about order of execution, it is ok for things to order differently in different runs.
Especially if it is hard to get around it.

> JavadocPackage 

Saying that, there will be some checks that can't have multiple instances in MT run, like: TranslationCheck, and JavadocPackageCheck . All threads must use the same instance to be able to provide accurate violations, especially for TranslationCheck as it is the one and only rare multi-file check. See Issue #3539 and #3540.

> should we replace TreeSet with HashSet and sort messages when all checks are finished,

If you are suggesting we store violations for all files that we scan until we are completely done, I don't think this is a good idea. The bigger the project, the more memory this will take and it is more likely users will get a memory exception before even seeing 1 violation.
IndentationCheck is a perfect example, as it is the check that produces the most violations. It is common for me to get out of memory exceptions on it during regression testing.

> we should fix compareTo method to follow the contract with equals.

Let's make a new issue to keep this thread on 1 track.

**Comment by soon** [comment]
> it is ok for things to order differently in different runs.

@rnveach Not sure, if I understand you correctly. The problem is not with the produced violations order. In the first run the checkstyle produces this error:

> [ERROR] /home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4861blockcommentstyle/InputCommentsIndentationSurroundingCode.java:0: Missing package-info.java file. [JavadocPackage]

And in the second run it produces another error:

> [ERROR] /home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4861blockcommentstyle/InputCommentsIndentationInSwitchBlock.java:0: Missing package-info.java file. [JavadocPackage]

Both these violations complains about the same directory  `/home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4861blockcommentstyle/` and therefore they are effectively the same. However, because they aren't strictly equal, I cannot just run checkstyle in MT mode twice and prepare a diff of violation reports, because they would definitely be different. 

This is why I'm asking to change the report format for this check, so that it looks like

> [ERROR] /home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4861blockcommentstyle: Missing package-info.java file. [JavadocPackage]

**Comment by rnveach** [comment]
> InputCommentsIndentationSurroundingCode.java
> InputCommentsIndentationInSwitchBlock.java

It is still the same, order of checks, order of files, order of violations.

>  I cannot just run checkstyle in MT mode twice and prepare a diff of violation reports, because they would definitely be different.

That is correct. MT mode will cause violations to be different depending on which checks or files gets run first. Thats why I was mentioning the other issue and why we don't care about order.

> This is why I'm asking to change the report format for this check

@romani Would have to agree, as all our violations have always been on files. This should probably also be in another issue.
This still doesn't change the fact that you may not be able to do this for all checks.
If this check isn't single instanced, it is still possible for you to get 2+ violations for the same directory as each thread could be working on different files with the same check in the same directory at the same time.

**Comment by soon** [comment]
> If this check isn't single instanced, it is still possible for you to get 2+ violations for the same directory as each thread could be working on different files with the same check in the same directory at the same time.

Sure. Now, this check isn't single instanced, but it holds ConcurrentHashMap, which is single instanced and passed through all checks. This, of course, will be changed as soon as I split all checks into groups (single instance / single instance per thread / single instance per file).

**Comment by MEZk** [comment]
> Let's make a new issue to keep this thread on 1 track.

Done. https://github.com/checkstyle/checkstyle/issues/4454

**Comment by romani** [comment]
Whole mess because multi-file Checks exists: TranslationCheck and JavadocPackage
Multi files validation is not supported in Checkstyle , if it is exists it is hack. We will not support this for long time.
Multi-file Checks should not have any annotation or have "Globalxxxxxx" annotation. Such Checks should be executed at the very end of Checkstyle as separate second execution with re-read of all files.
WE need to create internal two configs: with sinlge-file Checks and multi-file Checks. Filters should be shared the same. Merge of result(violations) should be done at the end.
ATTENTION: such allowance will be only for FilesetChecks which do file content examination them-self (no ANTRL parsing which is TreeWalker only).

As we remove such Checks from consideration....

So internally we will have two configs: for single file validation and multi-file validation.
Main thread will run multiple threads for single-file config , each thread will have it is own set of modules instances. So all modules will be completely tread local. In each thread(name="checkstyle-x") we run processing of whole file.
We will use ThreadPool and we will reuse Thread object to reuse tree of configured modules.
Ones we finish processing of all files by threads, .... control is in main thread again we run checkstyle again for multy-file Checks in main thread.
Printing results to user.

So as first step we will complete parallelization on the level of files, like running multiple chekcstyle on different sets of of files.

Ones we finish this, we can discuss MT mode in Treewalker(parsing, walikng, .....). But Treewalker MT will create threads inside "checkstyle-x" thread.

**Comment by romani** [comment]
#4890 is closed, but should be reviewed as it contains valuable discussion.
One day we will fins time to come back to MT.
